File: runme/mcgen/local/datan/solsrc/s2ds.f
      PROGRAM S2DS
      IMPLICIT DOUBLE PRECISION(A-H,O-Z)
      DIMENSION HIST(100),XPL(101),YPL(101)
      CHARACTER*75 TX,TY,CAPT
C identify program to user
      WRITE(*,*)' Program S2DS performs Monte Carlo folding '
      WRITE(*,*)' of uniform with Gaussian distribution'
      WRITE(*,*)' '
C ask for parameters
      WRITE(*,*)' Enter number NEPX of experiments (>>1)'
      WRITE(*,*)' >'
      READ(*,*)NEXP
      WRITE(*,*)' Enter left edge A of uniform distribution'
      WRITE(*,*)' >'
      READ(*,*)A
      WRITE(*,*)' Enter right edge B of uniform distribution'
      WRITE(*,*)' >'
      READ(*,*)B
      WRITE(*,*)'Enter width SIGMA of Gaussian distribution (>0.)'
      WRITE(*,*)' >'
      READ(*,*)SIGMA
C initialize histogram
      TX='x'
      TY='N(x)'
      LTX=1
      LTY=4
      CAPT(1:43)='n_exp#=*****, a=******, b=******, &s=******'
      WRITE(CAPT(8:12),'(I5)')NEXP
      WRITE(CAPT(17:22),'(F6.2)')A
      WRITE(CAPT(27:32),'(F6.2)')B
      WRITE(CAPT(38:43),'(F6.2)')SIGMA
      LCAPT=43
      H0=A-3.D0*SIGMA
      H1=B+3.D0*SIGMA
      DELH=.01D0*(H1-H0)
      NH=100
      CALL SMHSIN(HIST,H0,DELH,NH)
C loop over all experiments
      DO 30 IEXP=1,NEXP
        CALL RNECUY(RUNI,1)
        RUNI=A+(B-A)*RUNI
        CALL RNSTNR(RNORM,1)
        RNORM=RNORM*SIGMA
        ENTRY=RUNI+RNORM
C fill histogram histogram
        CALL SMHSFL(HIST,H0,DELH,NH,ENTRY,1.D0)
30    CONTINUE
C curve of folding between uniform and normal distribution
      NPL=NH+1
      FACT=DBLE(NEXP)*DELH/(B-A)
      DO 40 IPT=1,NPL
        XPL(IPT)=H0+(IPT-1)*DELH
        ARG1=(B-XPL(IPT))/SIGMA
        ARG2=(A-XPL(IPT))/SIGMA
        YPL(IPT)=FACT*(SCSTNR(ARG1)-SCSTNR(ARG2))
40    CONTINUE
C graphical output
      CALL GRNBWS()
      WRITE(*,*)' Please, enter number of workstation:'
      WRITE(*,*)' >'
      READ(*,*)NWS
      CALL GRHSCV(XPL,YPL,NPL,HIST,H0,DELH,NH,TX,LTX,TY,LTY,CAPT,LC
     +APT,NWS)
      END

File: runme/mcgen/local/datan/solsrc/s2ds.c
#include <stdio.h>
#include <string.h>
#include "datsrc.h"
#include "grsrc.h"

/* Table of constant values */

static integer c__1 = 1;
static double c_d1 = 1.;

main(void)
{
    /* Local variables */
    static integer iexp, lcapt, ltx, lty, nws, nexp, nh, npl, ipt;
    static double arg1, arg2, delh, fact, h0, h1, rnorm, entry__, hist[100],
                  runi, a, b, sigma, xpl[101], ypl[101];
    static char capt[75], tx[75], ty[75];
    /* System generated locals */
    integer i__1;
/* identify program to user */
    printf("%s\n", "Program S2DS performs Monte Carlo folding");
    printf("%s\n\n", "of uniform with Gaussian distribution.");
/* ask for parameters */
    printf("%s\n", "Enter number NEXP of experiments (>>1):");
    printf("%s", "> ");
    scanf("%li", &nexp);
    printf("%s\n", "Enter left edge A of uniform distribution:");
    printf("%s", "> ");
    scanf("%lf", &a);
    printf("%s\n", "Enter right edge B of uniform distribution:");
    printf("%s", "> ");
    scanf("%lf", &b);
    printf("%s\n", "Enter width SIGMA of Gaussian distribution (>0.):");
    printf("%s", "> ");
    scanf("%lf", &sigma);
/* initialize histogram */
    memset(tx, ' ', 75);
    memset(ty, ' ', 75);
    memset(capt, ' ', 75);
    strncpy(tx, "x", 1);
    strncpy(ty, "N(x)", 4);
    ltx = 1;
    lty = 4;
    sprintf(capt, "n_exp#=%5li, a=%6.2lf, b=%6.2lf, &s=%6.2lf",
            nexp, a, b, sigma);
    lcapt = strlen(capt);
    h0 = a - sigma * 3.;
    h1 = b + sigma * 3.;
    delh = (h1 - h0) * .01;
    nh = 100;
    smhsin_(hist, &h0, &delh, &nh);
/* loop over all experiments */
    i__1 = nexp;
    for (iexp = 1; iexp <= i__1; ++iexp) {
        rnecuy_(&runi, &c__1);
        runi = a + (b - a) * runi;
        rnstnr_(&rnorm, &c__1);
        rnorm *= sigma;
        entry__ = runi + rnorm;
/* fill histogram histogram */
        smhsfl_(hist, &h0, &delh, &nh, &entry__, &c_d1);
    }
/* curve of folding between uniform and normal distribution */
    npl = nh + 1;
    fact = (double) nexp * delh / (b - a);
    i__1 = npl;
    for (ipt = 1; ipt <= i__1; ++ipt) {
        xpl[ipt - 1] = h0 + (ipt - 1) * delh;
        arg1 = (b - xpl[ipt - 1]) / sigma;
        arg2 = (a - xpl[ipt - 1]) / sigma;
        ypl[ipt - 1] = fact * (scstnr_(&arg1) - scstnr_(&arg2));
    }
/* graphical output */
    grnbws_();
    printf("%s\n", "Please, enter number of workstation:");
    printf("%s", "> ");
    scanf("%li", &nws);
    (void)getchar();
    grhscv_(xpl, ypl, &npl, hist, &h0, &delh, &nh, tx, &ltx, ty, &lty, capt,
           &lcapt, &nws);
    return 0;
} /* main */

